yt/frontends/stream/data_structures.py

```diff
@@ -1092,11 +1138,71 @@ def __init__(self, stream_handler, storage_filename=None,
         if ('io', 'density') in fields and ('io', 'smoothing_length') in fields:
             self._sph_ptype = 'io'
 
-def load_particles(data, length_unit = None, bbox=None,
-                   sim_time=0.0, mass_unit = None, time_unit = None,
+    def add_sph_fields(self, n_neighbors=32, kernel="cubic", sph_ptype="io"):
+        """Add SPH fields for the specified particle type.
+
+        For a particle type with "particle_position" and "particle_mass" already
+        defined, this method adds the "smoothing_length" and "density" fields.
+        "smoothing_length" is computed as the distance to the nth nearest
+        neighbor. "density" is computed as the SPH (gather) smoothed mass.
+
+        Caution: this method will overwrite existing "smoothing_length" and
+        "density" fields.
```

Comment:
This is probably my last design question. Currently, the `add_sph_fields` method overwrites existing fields, which is the easiest to implement, less ambiguous to interpret, but not as flexible to use. My argument for this is that "smoothing_length" and "density" are usually computed together, so it doesn't make much sense to use this method to only generate one but not the other field. But I might have overlooked some use cases. Should I change this behavior?

Reply:
I think that's fine. Maybe it should noisily print a user-visible warning when a field actually gets clobbered?

We could also add a mode where we use user-supplied smoothing lengths or densities or both, and just be ok with the possibility that they aren't self-consistent.

Reply:
Okay, it's probably better to change this behavior then. I'll take a look. Thanks!

Reply:
In the new design, smoothing length and density are computed separately. And the `add_sph_fields` method only adds the SPH fields when they don't already exist.